mailpoet/lib/Util/Cookies.php

```diff
@@ -15,21 +15,39 @@ class Cookies {
 
   public function set($name, $value, array $options = []) {
     $options = $options + self::DEFAULT_OPTIONS;
-    $value = json_encode($value, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
+    $jsonValue = json_encode($value, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
     $error = json_last_error();
-    if ($error || ($value === false)) {
+    if ($error || ($jsonValue === false)) {
       throw new InvalidArgumentException();
     }
 
+
     // on PHP_VERSION_ID >= 70300 we'll be able to simply setcookie($name, $value, $options);
-    setcookie(
-      $name,
-      $value,
-      $options['expires'],
-      $options['path'],
-      $options['domain'],
-      $options['secure'],
-      $options['httponly']
+    if (!headers_sent()) {
+      setcookie(
+        $name,
+        $jsonValue,
+        $options['expires'],
+        $options['path'],
+        $options['domain'],
+        $options['secure'],
+        $options['httponly']
+      );
+      return;
+    }
+
+    $utcDate = (new \DateTime())->setTimestamp((int)$options['expires']);
+    $httpOnly = isset($options['httponly']) && $options['httponly'] ? '; httponly' : '';
+    $secure = isset($options['secure']) && $options['secure'] ? '; secure' : '';
+    printf(
+      "<script>document.cookie='%s=%s; expires=%s; path=%s; domain=%s%s%s'</script>",
+      esc_js($name),
+      wp_json_encode($value, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES),
+      esc_js($utcDate->format(\DATE_RFC1123)),
+      esc_js($options['path']),
+      esc_js($options['domain']),
+      esc_js($httpOnly),
+      esc_js($secure)
```

Comment:
@websupporter I see I'm assigned for review. I'm on L2 support so I shouldn't be but given that it's Friday afternoon, I can leave it for Monday anyway.

But I immediately have to ask — there's really no other solution to this? Such as setting the cookies earlier in the code that uses it? This looks quite hacky (or WordPressy 😆) to me. I really would expect a cookie service to set cookies and not print out some data. Do we really have no other choice? 😬

Reply:
So could we do `if (headers_sent()) { throw new SomeException(...); }` and make sure we never attempt that? I'm also worried about security and compatibility — for instance, what would `httpOnly` even mean in JS? Would it work? And if so, would it really be HTTP-only?

Reply:
> I'm on L2 support

Reply:
> We could hook into before_signup_header and check ourselfs, if its a user signup and whether the result of wpmu_validate_user_signup() is ok. That's a viable path.

I don't see into the details of it at all but if there's a viable path that would avoid the print, I'd say let's try to go for it!

> While I like the idea, simply throwing will likely throw exceptions in our integration test suite. But we can probably mock around it 🤔

Maybe it's not very safe in general. Instead of just some warnings in log files, it could break some sites where some plugins output stuff very early, etc. So I don't know, maybe safer to keep as-is?